test/libyul/YulOptimizerAssemblyTest.cpp

```
@@ -0,0 +1,106 @@
+#include <test/libyul/YulOptimizerTestCommon.h>
+
+#include <test/libyul/Common.h>
+#include <test/Common.h>
+
+#include <libyul/Object.h>
+#include <libyul/AsmPrinter.h>
+
+#include <liblangutil/SourceReferenceFormatter.h>
+#include <liblangutil/Scanner.h>
+
+#include <libsolutil/AnsiColorized.h>
+#include <libsolutil/CommonIO.h>
+
+#include <libyul/YulStack.h>
+#include <libyul/AsmAnalysis.h>
+#include <libyul/AsmAnalysisInfo.h>
+#include <libyul/backends/evm/EVMDialect.h>
+#include <libyul/backends/evm/EthAssemblyAdapter.h>
+#include <libyul/backends/evm/EVMObjectCompiler.h>
+
+#include <libevmasm/Assembly.h>
+
+#include <fstream>
+
+using namespace solidity;
+using namespace solidity::util;
+using namespace solidity::langutil;
+using namespace solidity::yul;
+using namespace solidity::yul::test;
+using namespace solidity::frontend;
+using namespace solidity::frontend::test;
+
+YulOptimizerAssemblyTest::YulOptimizerAssemblyTest(std::string const& _filename):
+	EVMVersionRestrictedTestCase(_filename)
+{
+	boost::filesystem::path path(_filename);
+
+	if (path.empty() || std::next(path.begin()) == path.end() || std::next(std::next(path.begin())) == path.end())
+		BOOST_THROW_EXCEPTION(std::runtime_error("Filename path has to contain a directory: \"" + _filename + "\"."));
+	m_optimizerStep = std::prev(std::prev(path.end()))->string();
+
+	m_source = m_reader.source();
+	m_expectation = m_reader.simpleExpectations();
+}
+
+TestCase::TestResult YulOptimizerAssemblyTest::run(std::ostream& _stream, std::string const& _linePrefix, bool const _formatted)
+{
+	std::shared_ptr<Object> object;
+	std::shared_ptr<AsmAnalysisInfo> analysisInfo;
+	Parser::DebugAttributeCache::Ptr debugAttributeCache = std::make_shared<Parser::DebugAttributeCache>();
+
+	Dialect const& dialect = test::dialect("evm", solidity::test::CommonOptions::get().evmVersion());
+	std::tie(object, analysisInfo) = test::parse(_stream, _linePrefix, _formatted, m_source, dialect, debugAttributeCache);
```

Just a heads-up: while reviewing #15258 today I realized that even though the parsing here will accept full Yul objects with subobjects, `YulOptimizerTestCommon` only really works on the top-level AST (see https://github.com/ethereum/solidity/pull/15258#discussion_r1700298830), so the output will not really match what solc would output when optimizing a full object, which is highly misleading. There I recommended making `YulOptimizerTestCommon` assert against such objects as a quick fix.

This means that if you do need to test debug data on full objects, you'll have to extend `YulOptimizerTestCommon` to actually support them properly. If you don't, you should probably instead raise an error here not to run into the proposed assertion.